### ISSUE TITLE
NULL pointer exception, version v4.3 stable

### ISSUE THREAD

**Issue by windflowerly** [warmcat/libwebsockets#3099]
![image](https://github.com/warmcat/libwebsockets/assets/3378243/7db7c8a1-83a6-4670-9d1d-7a4d7cb143cd)

stack:
[0] from 0x00007f302e4a0e0d in __lws_sul_service_ripe+254 at ./libwebsocket/lib/core-net/sorted-usec-list.c:137
[1] from 0x00007f302e49fbf7 in lws_service_adjust_timeout+251 at ./libwebsocket/lib/core-net/service.c:323
[2] from 0x00007f302e6f77db in lws_uv_idle+143 at ./libwebsocket/lib/event-libs/libuv/libuv.c:73
[3] from 0x00000000004859f4 in uv.run_idle
[4] from 0x00000000004837ce in uv_run
[5] from 0x00007f302e6f8fb5 in elops_run_pt_uv+111 at ./libwebsocket/lib/event-libs/libuv/libuv.c:647
[6] from 0x000000000043a813 in lws_service_tsi+158 at ./libwebsocket/lib/core-net/service.c:863
[7] from 0x0000000000421435 in thread_service(void*)+75 at ./xxx.cpp:64
[8] from 0x00007f3031625ea5 in start_thread

**Comment by lws-team** [comment]
This is the signature of not 'cancelling' a sul before destroying the object containing the sul.  Cancelling the sul before destroying its storage makes sure it is removed from any linked-lists beforehand.

See https://github.com/warmcat/libwebsockets/blob/e4be3317ee421f951fb16a85c7edacc9b8a0e6aa/include/libwebsockets/lws-timeout-timer.h#L208

**Comment by windflowerly** [comment]
but i never called lws_sul2_schedule functioin.
Just make 10K ws client connections。

**Comment by lws-team** [comment]
Notice lws is not threadsafe...

**Comment by windflowerly** [comment]
all action in lws thread context

**Comment by lws-team** [comment]
... you seem to spawn a thread though, why not try without that.

Still problems on main branch?  If so, send a reproducer.

**Comment by windflowerly** [comment]
I try it.

**Comment by windflowerly** [comment]
A build error on main branch:
n file included from ./libwebsocket/lib/plat/unix/unix-sockets.c:31:
/usr/include/linux/ipv6.h:19:8: error: redefinition of ‘struct in6_pktinfo’
 struct in6_pktinfo {
        ^~~~~~~~~~~
In file included from ./libwebsocket/lib/plat/unix/./private-lib-plat-unix.h:31,
                 from ./libwebsocket/lib/core/./private-lib-core.h:138,
                 from ./libwebsocket/lib/plat/unix/unix-sockets.c:28:
/usr/include/netinet/in.h:538:8: note: originally defined here
 struct in6_pktinfo
        ^~~~~~~~~~~
In file included from ./libwebsocket/lib/plat/unix/unix-sockets.c:31:
/usr/include/linux/ipv6.h:24:8: error: redefinition of ‘struct ip6_mtuinfo’
 struct ip6_mtuinfo {
        ^~~~~~~~~~~
In file included from ./libwebsocket/lib/plat/unix/./private-lib-plat-unix.h:31,
                 from ./libwebsocket/lib/core/./private-lib-core.h:138,
                 from ./libwebsocket/lib/plat/unix/unix-sockets.c:28:
/usr/include/netinet/in.h:545:8: note: originally defined here
 struct ip6_mtuinfo
        ^~~~~~~~~~~


![image](https://github.com/warmcat/libwebsockets/assets/3378243/441a8efa-5b4b-4ae9-98d7-15f099166ad0)

**Comment by lws-team** [comment]
... what is your platform?  This doesn't occur on recent Linux + glibc.

**Comment by windflowerly** [comment]
CentOS Linux release 7.9.2009 (Core)

[root@localhost build]# ldd --version
ldd (GNU libc) 2.17

Linux localhost.localdomain 3.10.0-1160.102.1.el7.x86_64 #1 SMP Tue Oct 17 15:42:21 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

**Comment by windflowerly** [comment]
minimal-ws-server-threads-foreign-libuv-smp will reproduce.
1. use jmeter make 10K ws connection to websocket server
2. kill jmeter ( must  kill jmeter )

**Comment by lws-team** [comment]
Eg, Rocky 9.3 doesn't have this problem.

Looking in the libc headers, it seems those structs are defined in two places, there's some magic in /usr/include/linux/libc-compat.h that selects between them.

**Comment by windflowerly** [comment]
When the number of connections is large, if there is an abnormal disconnection, LWS(v4.3-stable) is unstable

**Comment by windflowerly** [comment]
> Eg, Rocky 9.3 doesn't have this problem.
> 
> Looking in the libc headers, it seems those structs are defined in two places, there's some magic in /usr/include/linux/libc-compat.h that selects between them.

only one place define the structs

![image](https://github.com/warmcat/libwebsockets/assets/3378243/1f052686-d605-417e-b60c-4d0af8996607)

**Comment by lws-team** [comment]
Your grep isn't right to find both.

Eg, on Rocky 9.3

```
$ grep in6_pktinfo /usr/include -R
/usr/include/netinet/in.h:struct in6_pktinfo
/usr/include/linux/ipv6.h:struct in6_pktinfo {
```

```
$ rpm -q glibc
glibc-2.34-83.el9.12.x86_64
```

**Comment by windflowerly** [comment]
sorry.

[root@localhost build]# rpm -q glibc
glibc-2.17-326.el7_9.x86_64

**Comment by windflowerly** [comment]
has this issue reproduce?
It is likely that multiple service mode will trigger this issue.

**Comment by windflowerly** [comment]
hi, did you reproduce the null pointer exception ?

**Comment by lws-team** [comment]
Somebody found the root cause for this recently, the fix should be on main + v4.3-stable.
